src/block/Hopper.php

```diff
@@ -96,8 +103,238 @@ public function onInteract(Item $item, int $face, Vector3 $clickVector, ?Player
 	}
 
 	public function onScheduledUpdate() : void{
-		//TODO
+		$this->position->getWorld()->scheduleDelayedBlockUpdate($this->position, 1);
+
+		$tile = $this->position->getWorld()->getTile($this->position);
+		if(!$tile instanceof TileHopper){
+			return;
+		}
+
+		$transferCooldown = $tile->getTransferCooldown();
+		if($transferCooldown > 0){
+			$transferCooldown--;
+			$tile->setTransferCooldown($transferCooldown);
+		}
+
+		if($this->isPowered()){
+			// If a hopper is powered it is deactivated and won't push, pull or pick up items.
+			return;
+		}
+
+		if($transferCooldown > 0){
+			// Skipping this tick because the hopper is still on cooldown of 8 ticks.
+			return;
+		}
+
+		$success = $this->push($tile);
+		// Hoppers that have a container above them, won't try to pick up items.
+		$origin = $this->position->getWorld()->getTile($this->position->getSide(Facing::UP));
+		if($origin instanceof Container){
+			$success |= $this->pull($tile, $origin);
+		}else{
+			$success |= $this->pickup($tile);
+		}
+		// The cooldown is only set back to 8 ticks if the hopper has done anything.
+		if((bool) $success){
```

Comment:
This cast seems unnecessary given the fact that Hopper::push(), Hopper::pull(), and Hopper::pickup() all return booleans.

Reply:
The bitwise operations cast $success at this point to an integer of 0 (false) or 1 (true).
I know that it is automatically cast to a boolean in the if statement but I made this to ensure that this definitely is a boolean and won't break in the future (idk.. if `declare(strict_types=1)` would stop auto casting in if statements in a future php version)

Reply:
The bitwise operators were nice because they helped me implement the hopper logic (only reset cooldown if done anything) with just these few lines.
And I choose to cast it back to a boolean instead of comparing its value because it looked more clear to me.